src/search-and-replace-dialog.jsx

```diff
@@ -0,0 +1,286 @@
+        }
+
+        this.props.onChange({ searchIndex });
+    },
+
+    handleReplaceAll() {
+        var replaceString = this.state.replaceString;
+        var { searchString, question, hints, particle } = this.props;
+
+        var regex = new RegExp(Util.escapeRegExp(searchString), "g");
+
+        var replaceFunc = function(obj) {
+            if (obj) {
+                if (Array.isArray(obj)) {
+                    obj.forEach(replaceFunc);
+                } else {
+                    // excludes occurrences within widget references
+                    var indices = Util.getIndicesOf(obj.content, searchString);
+                    obj.content = obj.content.replace(regex, (match, offset) => {
+                        // make sure this match isn't inside a widget reference
+                        if (indices.indexOf(offset) !== -1) {
+                            return replaceString
+                        } else {
+                            return match;
+                        }
+                    });
+                }
+            }
+        };
+
+        replaceFunc(question);
+        replaceFunc(hints);
+        replaceFunc(particle);
+
+        // handle the case where the replaceString contains one or more copies
+        // of searchString
+        // TODO(kevinb7) this is quite right
+        var searchResultCount = this.getSearchResultCount(searchString);
+
+        this.setState({ searchResultCount });
+        this.props.onChange({ question, hints, json: particle, searchIndex: 0 });
+    },
+
+    handleReplace() {
+        var replaceString = this.state.replaceString;
+        var { searchIndex, searchString, question, hints, particle } = this.props;
+
+        var regex = new RegExp(searchString, "g");
+        var replaced = false;
+        var globalIndex = 0;
+
+        var replaceFunc = function(obj) {
+            if (!replaced && obj) {
+                if (Array.isArray(obj)) {
+                    obj.forEach(replaceFunc);
+                } else {
+                    // excludes occurrences within widget references
+                    var indices = Util.getIndicesOf(obj.content, searchString);
+                    obj.content = obj.content.replace(regex, (match, offset) => {
+                        // make sure this match isn't inside a widget reference
+                        if (indices.indexOf(offset) !== -1) {
+                            if (!replaced && globalIndex === searchIndex) {
+                                replaced = true;
+                                globalIndex++;
+                                return replaceString;
+                            } else {
+                                globalIndex++;
+                                return match;
+                            }
+                        } else {
+                            return match;
+                        }
+                    });
+                }
+            }
+        };
```

Comment:
this looks really similar to `handleReplaceAll`. Maybe we can factor out the common pieces?